File: qc-application/src/main/java/com/eager/questioncloud/application/api/subscribe/implement/SubscribeProcessor.java
package com.eager.questioncloud.application.api.subscribe.implement;

import com.eager.questioncloud.core.domain.creator.infrastructure.repository.CreatorRepository;
import com.eager.questioncloud.core.domain.subscribe.infrastructure.SubscribeRepository;
import com.eager.questioncloud.core.domain.subscribe.model.Subscribe;
import com.eager.questioncloud.core.exception.CoreException;
import com.eager.questioncloud.core.exception.Error;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class SubscribeProcessor {
    private final SubscribeRepository subscribeRepository;
    private final CreatorRepository creatorRepository;

    public Subscribe subscribe(Long userId, Long creatorId) {
        if (!isActiveCreator(creatorId)) {
            throw new CoreException(Error.NOT_FOUND);
        }

        if (isAlreadySubscribed(userId, creatorId)) {
            throw new CoreException(Error.ALREADY_SUBSCRIBE_CREATOR);
        }

        return subscribeRepository.save(Subscribe.create(userId, creatorId));
    }

    public void unSubscribe(Long userId, Long creatorId) {
        subscribeRepository.unSubscribe(userId, creatorId);
    }

    private Boolean isActiveCreator(Long creatorId) {
        return creatorRepository.existsById(creatorId);
    }

    private Boolean isAlreadySubscribed(Long userId, Long creatorId) {
        return subscribeRepository.isSubscribed(userId, creatorId);
    }
}

File: qc-application/src/main/kotlin/com/eager/questioncloud/application/api/subscribe/implement/SubscribeProcessor.kt
package com.eager.questioncloud.application.api.subscribe.implement

import com.eager.questioncloud.core.domain.creator.infrastructure.repository.CreatorRepository
import com.eager.questioncloud.core.domain.subscribe.infrastructure.repository.SubscribeRepository
import com.eager.questioncloud.core.domain.subscribe.model.Subscribe.Companion.create
import com.eager.questioncloud.core.exception.CoreException
import com.eager.questioncloud.core.exception.Error
import org.springframework.stereotype.Component

@Component
class SubscribeProcessor(
    private val subscribeRepository: SubscribeRepository,
    private val creatorRepository: CreatorRepository,
) {
    fun subscribe(userId: Long, creatorId: Long) {
        if (!isActiveCreator(creatorId)) {
            throw CoreException(Error.NOT_FOUND)
        }

        if (isAlreadySubscribed(userId, creatorId)) {
            throw CoreException(Error.ALREADY_SUBSCRIBE_CREATOR)
        }

        subscribeRepository.save(create(userId, creatorId))
    }

    fun unSubscribe(userId: Long, creatorId: Long) {
        subscribeRepository.unSubscribe(userId, creatorId)
    }

    private fun isActiveCreator(creatorId: Long): Boolean {
        return creatorRepository.existsById(creatorId)
    }

    private fun isAlreadySubscribed(userId: Long, creatorId: Long): Boolean {
        return subscribeRepository.isSubscribed(userId, creatorId)
    }
}
